File: GO_ogle/MoveZeros.py
class Solution(object):
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        if not nums:
            return nums
        for i in range(1, len(nums)):
            if nums[i - 1] == 0 and nums[i] != 0:
                j = i - 1
                while j > 0 and nums[j - 1] == 0:
                        j -= 1
                nums[j], nums[i] = nums[i], nums[j]
        return nums


solution = Solution()
print(solution.moveZeroes([0, 1, 0, 3, 12]))

File: FaceBook/MoveZeros.scala
object Solution extends App {
  def moveZeroes(nums: Array[Int]): Unit = {
    var z = 0

    for (i <- 0 until nums.length) {
      if (nums(i) != 0) {
        val numsI = nums(i)
        nums(i) = nums(z)
        nums(z) = numsI
        z += 1
      }
    }
  }
}

/*
G: nums: Array[Int]
O: Unit
T: O(n)
S: O(1)

Notes:
  - Move zeros to the end of the array
  - Maintain order of non zero elements
  - Do it in place
  - Minimize number of ops

Ex1:
[0,1,0,3,12] => [1,3,12,0,0]

1 3 0 0 12
    i
        j

 */
